File: csharp/matching-brackets/MatchingBrackets.cs
using System;
using System.Collections.Generic;

public static class MatchingBrackets
{
    public static bool IsPaired(string input)
    {
        var opened = new List<char>{ '(', '[', '{' };
        var closed = new List<char>{ ')', ']', '}' };
        var brackets = new Stack<char>();
        for (int i = 0; i < input.Length; i++)
        {
            if (!System.Char.IsPunctuation(input[i])) continue;
            if (opened.Contains(input[i])) brackets.Push(input[i]);
            if (closed.Contains(input[i]))
            {
                if (brackets.Count == 0) return false;
                char openBracket = brackets.Pop();
                int openBracketIndex = opened.FindIndex(c => c == openBracket);
                if (openBracketIndex < 0) return false;
                int closedBracketIndex = closed.FindIndex(c => c == input[i]);
                if (openBracketIndex != closedBracketIndex) return false;
            }
        }
        if (brackets.Count != 0)
            return false;
        return true;
    }
}

File: fsharp/matching-brackets/MatchingBrackets.fs
module MatchingBrackets

let fillBracketStack (l: char list option) (c: char): char list option =
    let opened = ['(';'[';'{']
    let closed = [')';']';'}']
    if List.contains c opened then 
        Option.map (fun ls -> c :: ls) l
    elif List.contains c closed then
        match l with 
        | None -> None
        | Some [] -> None
        | Some (hd :: tl) -> 
            if List.findIndex ((=) c) closed = List.findIndex ((=) hd) opened then 
                Some tl 
            else 
                None
    else
        l

let isPaired (input: string): bool = 
    match Seq.fold fillBracketStack (Some []) input with
    | None -> false
    | Some ls -> ls |> List.isEmpty